### PR TITLE
use float 16 in causal mask and masked bias

### LINKED ISSUE #13193: Megatron conversion code converts some weights in fp16 to fp32(or uint8).
## Environment info

transformers version: 4.9.2
Platform: Linux-4.18.0-25-generic-x86_64-with-glibc2.10
Python version: 3.8.5
PyTorch version (GPU?): 1.8.0a0+52ea372 (True)
Tensorflow version (GPU?): not installed (NA)
Flax version (CPU?/GPU?/TPU?): not installed (NA)
Jax version: not installed
JaxLib version: not installed
Using GPU in script?:
Using distributed or parallel set-up in script?: No

### Who can help

@novatig @jdemouth @LysandreJik

## Information

Model I am using (Bert, XLNet ...):

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [ ] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [ ] my own task or dataset: (give details below)

## To reproduce

1. Check the data type of original megatron checkpoint. It's all in fp16.
```bash
wget --content-disposition  https://api.ngc.nvidia.com/v2/models/nvidia/megatron_lm_345m/versions/v0.0/zip -O checkpoint.zip
unzip checkpoint.zip
python -c "import torch; from pprint import pprint as print; sd=torch.load('./release/mp_rank_00/model_optim_rng.pt'); d= {d.dtype: 1 for d in sd['model']['language_model']['transformer'].values()}; print(d.keys())"
# dict_keys([torch.float16])
```

2. But the [current conversion script](https://github.com/huggingface/transformers/blob/master/src/transformers/models/megatron_gpt2/convert_megatron_gpt2_checkpoint.py) converts some into [float32](https://github.com/huggingface/transformers/blob/91ff480e2693f36b11aaebc4e9cc79e4e3c049da/src/transformers/models/megatron_gpt2/convert_megatron_gpt2_checkpoint.py#L164) and [uint8](https://github.com/huggingface/transformers/blob/91ff480e2693f36b11aaebc4e9cc79e4e3c049da/src/transformers/models/megatron_gpt2/convert_megatron_gpt2_checkpoint.py#L160). This leads to a model with data type which is not faithful to the original model, and potentially a problem as discussed in #13076 
```
python3 /hf/transformers-master/src/transformers/models/megatron_bert/convert_megatron_gpt2_checkpoint.py checkpoint.zip
python -c "import torch; sd=torch.load('pytorch_model.bin');  d = {p.dtype:1 for p in sd.values() }; print(d.keys())"
# dict_keys([torch.float16, torch.float32, torch.uint8])
```

## Expected behavior
Converted checkpoint should have the same data type as the original one.

I will open a new PR to address this :)

### PR DESCRIPTION
**Pull request by hwijeen**
# What does this PR do?

Fixes #13193  (issue). Problem description and reproducer is provided in the issue.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@LysandreJik @NielsRogge as they reviewed [the original converting script PR](https://github.com/huggingface/transformers/pull/12007)

### CODE REVIEW

**Comment by LysandreJik** [comment]
Pinging @jdemouth and @novatig

**Comment by jdemouth** [comment]
@novatig it looks good to me. Are you ok with the changes? 

@hwijeen and @LysandreJik, sorry for the delay, I was on holidays ;)

**Comment by LysandreJik** [comment]
No worries, thanks for taking a look! :)

**Comment by hwijeen** [comment]
This is a kindly reminder for @novatig :)

**Comment by LysandreJik** [comment]
Merging since @jdemouth approved - will reverse if @novatig disagrees.

**Comment by novatig** [comment]
Sorry all, I did not see the notification in my inbox and it slipped my mind.

A very belated LGTM
